File: src-tauri/src/timer.rs
use std::time::Duration;

use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Manager};
use tokio::{sync::mpsc::Receiver, task::AbortHandle, time::interval};

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TickPayload {
    new_value: u32,
}

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CancelPayload {
    reason: CancelReason,
}

#[derive(Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum CancelReason {
    Cancelled,
    Finished,
    Paused,
}

///  Starts a timer. This exposes 2 events:
///  1. tickTimer: Invoked every time the tick timer changes.
///  2. cancelTimer: Emitted (or received) when the timer is cancelled.
#[tauri::command]
pub async fn start_timer(duration: u32, app_handle: AppHandle) {
    // Sleeping for 1s fixes so many weird bugs. It works, don't touch it.
    // tokio::time::sleep(Duration::from_secs(1)).await;
    let (mut rx, abort_handle) = schedule_timer(duration);
    app_handle.once_global("cancelTimer", move |_event| {
        abort_handle.abort();
        // rx.close();
    });
    tokio::spawn(async move {
        while let Some(value) = rx.recv().await {
            app_handle
                .emit_all("tickTimer", TickPayload { new_value: value })
                .unwrap();
            if value <= 0 {
                app_handle
                    .emit_all(
                        "cancelTimer",
                        CancelPayload {
                            reason: CancelReason::Finished,
                        },
                    )
                    .unwrap();
            }
        }
    });
}

fn schedule_timer(seconds: u32) -> (Receiver<u32>, AbortHandle) {
    let (tx, rx) = tokio::sync::mpsc::channel::<u32>(100);
    let handle = tokio::spawn(async move {
        let mut value = seconds;
        let mut interval = interval(Duration::from_secs(1));
        interval.tick().await;
        tx.send(value).await.unwrap();
        loop {
            interval.tick().await;
            value = value.wrapping_sub(1);
            tx.send(value).await.unwrap();
        }
    });
    return (rx, handle.abort_handle());
}

File: src/bindings/timer.ts
import { invoke } from "@tauri-apps/api";
import { emit, listen, once } from "@tauri-apps/api/event";

/**
 * The reason the timer was cancelled.
 */
export type CancelReason = "cancelled" | "finished" | "paused";

/**
 * The function to stop the timer
 */
export type StopFn = (reason: CancelReason) => void;

export interface CancelTimerPayload {
    reason: CancelReason;
}

export interface TickTimerPayload {
    newValue: number;
}

/**
 * Starts the timer.
 *
 * @param seconds Timer duration
 * @param onTick invoked on every second
 * @param onCancelled Invoked when the timer is cancelled.
 * @returns The function to cancel the timer.
 */
export async function startTimer(
    seconds: number,
    onTick: (value: number) => void,
    onCancelled: (reason: CancelReason) => void
): Promise<StopFn> {
    await invoke("start_timer", { duration: seconds });
    const stopFn = (reason: CancelReason) => {
        emit("cancelTimer", { reason });
    };
    const unlistenToTick = await listen<TickTimerPayload>(
        "tickTimer",
        (event) => {

            onTick(event.payload.newValue as number);
        }
    );

    await once<CancelTimerPayload>("cancelTimer", (event) => {
        unlistenToTick();
        onCancelled(event.payload.reason);
    });
    return stopFn;
}
